core/src/main/java/hudson/PluginManager.java

```diff
@@ -811,9 +838,36 @@ public void doInstall(StaplerRequest req, StaplerResponse rsp) throws IOExceptio
                 if (p == null) {
                     throw new Failure("No such plugin: " + n);
                 }
-                p.deploy(dynamicLoad);
+                
+                deployJobs.add(p.deploy(dynamicLoad));
             }
         }
+        
+        // Fire a one-off thread to wait for the plugins to be deployed and then
+        // refresh the dependant plugins list.
+        new Thread() {
+            @Override
+            public void run() {
+                INSTALLING: while (true) {
+                    for (Future<UpdateCenter.UpdateCenterJob> deployJob : deployJobs) {
+                        try {
+                            Thread.sleep(500);
+                        } catch (InterruptedException e) {
+                            LOGGER.log(SEVERE, "Unexpected error while waiting for some plugins to install. Plugin Manager state may be invalid. Please restart Jenkins ASAP.", e);
+                        }
+                        if (!deployJob.isCancelled() && !deployJob.isDone()) {
+                            // One of the plugins is not installing/canceled, so
+                            // go back to sleep and try again in a while.
+                            continue INSTALLING;
+                        }
+                    }
+                    // All the plugins are installed. It's now safe to refresh.
+                    resolveDependantPlugins();
+                    break;
+                }
+            }
+        }.start();
+        
```

Comment:
@batmat good question. My thinking in this particular case was to keep it simple. This thread only gets created when you install plugins, which would be a relatively low frequency thing. As for stopping it on shutdown, I was sure the plugin deploy jobs get aborted on a shutdown, which would cause this thread to exit at the same time. So, it seemed like overkill to implement this in any other way.